framework/include/parser/CommandLine.h

```diff
@@ -116,190 +169,190 @@ class CommandLine
+      // Keep track of and change the throw on error characteristics so that
+      // we can catch parsing errors for the argument
+      const auto throw_on_error_orig = Moose::_throw_on_error;
+      Moose::_throw_on_error = true;
 
-template <typename T>
-bool
-CommandLine::search(const std::string & option_name, T & argument)
-{
-  if (auto pos = _cli_options.find(option_name); pos != _cli_options.end())
-  {
-    const auto & option = pos->second;
-    for (const auto & cli_switch : option.cli_switch)
-      for (const auto arg_i : index_range(_args))
+      try
       {
-        const auto & arg = _args[arg_i];
-
-        if (arg == cli_switch)
-        {
-          // "Flag" CLI options are added as Boolean types, when we see them
-          // we set the Boolean argument to true
-          if (option.argument_type == ArgumentType::NONE)
-            argument = true;
-          else if (arg_i + 1 < _args.size())
-          {
-            std::stringstream ss;
-            ss << _args[arg_i + 1];
-
-            setArgument(ss, argument, cli_switch);
-          }
-          else if (option.argument_type == ArgumentType::REQUIRED)
-          {
-            mooseError("The command line argument '",
-                       cli_switch,
-                       "' requires a value and one was not provided.");
-          }
-          return true;
-        }
+        if constexpr (std::is_same_v<type, std::string> || std::is_same_v<type, MooseEnum>)
+          value = from;
+        else
+          value = MooseUtils::convert<type>(from, true);
+      }
+      catch (std::exception & e)
+      {
+        Moose::_throw_on_error = throw_on_error_orig;
+        mooseError("While parsing command line option '",
+                   cli_switch,
+                   "' with value '",
+                   from,
+                   "':\n\n",
+                   e.what());
       }
 
-    if (pos->second.required)
-    {
-      Moose::err << "Required parameter: " << option_name << " missing\n";
-      printUsage();
-    }
-    return false;
-  }
-  mooseError("Unrecognized option name");
-}
-
-template <typename T>
-bool
-CommandLine::search(const std::string & option_name, std::vector<T> & argument)
-{
-  std::map<std::string, Option>::iterator pos = _cli_options.find(option_name);
-  if (pos != _cli_options.end())
-  {
-    for (unsigned int i = 0; i < pos->second.cli_switch.size(); ++i)
+      Moose::_throw_on_error = throw_on_error_orig;
+    };
+
+    // If a value doesn't exist, check the next argument to see if it
+    // would work. This is needed for when we have argument values that
+    // have = signs that get split. For example:
+    // "--required-capabilities 'petsc>=3.11'" would get split into:
+    //   - "--required-capabilities" with no value
+    //   - "petsc>" with value "3.11"
+    // which we want to re-combine into
+    //   - "--required-capabilities" with value "petsc>=3.11"
```

Comment:
Aren't you also splitting along spaces? Is
`--required-capabilities 'petsc>=3.11 & ad_size=30 & libtorch'"`

maybe resulting in a valid "HIT parameter" `ad_size=30` being parsed accidentally?

Reply:
Tested that this works as expected